wgpu/src/lib.rs

```diff
@@ -632,6 +632,57 @@ impl Drop for ShaderModule {
     }
 }
 
+impl ShaderModule {
+    /// Get the compilation info for the shader module.
+    pub async fn compilation_info(&self) -> CompilationInfo {
+        todo!()
+    }
+}
+
+/// Compilation information for a shader module.
+///
+/// Corresponds to [WebGPU `GPUCompilationInfo`](https://gpuweb.github.io/gpuweb/#gpucompilationinfo).
+/// The source locations use bytes, and index a UTF-8 encoded string.
+#[derive(Debug)]
+pub struct CompilationInfo {
+    /// The messages from the shader compilation process.
+    pub messages: Vec<CompilationMessage>,
+}
+
+/// A single message from the shader compilation process.
+///
+/// Roughly corresponds to [`GPUCompilationMessage`][https://www.w3.org/TR/webgpu/#gpucompilationmessage],
+/// except that there is only a `span`, with a byte positions in a UTF-8 encoded source code.
+#[derive(Debug)]
+pub struct CompilationMessage {
+    /// The text of the message.
+    pub message: String,
+    /// The type of the message.
+    pub message_type: CompilationMessageType,
+    /// Where in the source code the message points at.
+    pub span: Option<SourceSpan>,
+}
+
+/// The type of a compilation message.
+#[derive(Debug)]
+pub enum CompilationMessageType {
+    /// An error message.
+    Error,
+    /// A warning message.
+    Warning,
+    /// An informational message.
+    Info,
+}
+
+/// Byte positions in an UTF-8 encoded string.
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+pub struct SourceSpan {
+    /// 0-based offset in code units (in bytes) of the start of the span.
+    pub offset: u32,
+    /// Length in code units (in bytes) of the span.
+    pub length: u32,
+}
```

Comment:
Was creating a new type here the correct move, or should I just have exposed the already existing Naga `Span` type?

Also, the WebGPU spec wants a `lineNum` and a `linePos`. Should I add those to this struct and thus redundantly store them? Or add a helper method to compute them? Or neither of those options?